Permissive slug generation, refs #11761

sbreker opened the pull request:
Ensure the characters * : @ = , are represented as literals in the
browser address bar when the 'Permissive Slugs' setting is activated.

- this will prevent these chars from being percent-encoded when
  'permissive slugs' setting is turned on in AtoM
- these chars are allowed in slugs when 'permissive slugs' setting is on.
- if 'permissive slugs' is OFF, use PHP's urlencode as normal.
- this should only affect the PATH segment of an AtoM URI.
- this will affect URI construction for page links and address bar.

sbreker committed:
Permissive slug generation, refs #11761

Ensure the characters * : @ = , are represented as literals in the
browser address bar when the 'Permissive Slugs' setting is activated.

- this will prevent these chars from being percent-encoded when
  'permissive slugs' setting is turned on in AtoM
- these chars are allowed in slugs when 'permissive slugs' setting is on.
- if 'permissive slugs' is OFF, use PHP's urlencode as normal.
- this should only affect the PATH segment of an AtoM URI.
- this will affect URI construction for page links and address bar.

jraddaoui commented on vendor/symfony/lib/routing/sfRoute.class.php:
```diff
@@ -283,7 +283,7 @@ protected function generateWithTokens($parameters)
           if (!$optional || !isset($this->defaults[$token[3]]) || $parameters[$token[3]] != $this->defaults[$token[3]])
           {
-            $url[] = urlencode($parameters[$token[3]]);
+            $url[] = QubitRoute::urlencode3986($parameters[$token[3]]);
```
We should try to not make changes in the Symfony classes when possible. I'd overwrite the entire function in the QubitRoute class. I'm not sure if all the other QubitRoute classes extend QubitRoute or sfRoute, so you may need to do the same in others or add an intermediate class.

jraddaoui commented on lib/routing/QubitRoute.class.php:
```diff
@@ -51,6 +51,25 @@ protected function filterParams($params)
+  * of a URI - this is the path segment.
+  *  - these chars are allowed in slugs when 'permissive slugs' setting is on.
+  *  - if 'permissive slugs' is OFF, use urlencode as normal.
+  *  - this should ONLY affect the slug portion of an AtoM URI
```
Are all those hyphens intended? I'd prefer a normal paragraph ;)

jraddaoui reviewed (changes requested):
Looks great @sbreker!

But I'd try to avoid making changes in the Symfony classes if it's possible.

sbreker committed:
Permissive Slug Generation fix, refs #11761

Removed modification to sfRoute and added code to override method in
QubitRoute. QubitResourceRoute is already extending QubitRoute.

I do not believe QubitMetadataRoute needs to be changed as it appears
to be related to request routing which will correctly resolve requests
with either literals or percent-encoded unicode characters in them.

In testing I also found an issue with generating finding aids when a
slug has a semicolon in it - in the exec call to run export:bulk, the
slug needs to be enclosed in quotes to ensure characters in the slug
are not interpreted on the command line.

jraddaoui commented on lib/routing/QubitRoute.class.php:
```diff
@@ -85,4 +85,58 @@ public function generate($params, $context = array(), $absolute = false)
+          }
+          break;
+        default:
+          // handle custom tokens
```
Uppercase.

jraddaoui reviewed (approved):
Looks great @sbreker! Just a really minor comment.
